src/Autofac.Configuration/Util/TypeManipulation.cs

```diff
@@ -144,6 +144,11 @@ public static object ChangeToCompatibleType(object value, Type destinationType,
             converter = TypeDescriptor.GetConverter(destinationType);
             if (converter.CanConvertFrom(value.GetType()))
             {
+                if (converter is BaseNumberConverter)
+                {
+                    value = value.ToString().ReplaceDecimalSeparator();
+                }
+
                 return converter.ConvertFrom(value);
```

Comment:
I think this is where we'd need to [specify the invariant culture for the conversion](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.typeconverter.convertfrom?view=netframework-4.8#System_ComponentModel_TypeConverter_ConvertFrom_System_ComponentModel_ITypeDescriptorContext_System_Globalization_CultureInfo_System_Object_)

Reply:
Yes, I tried that, didn't worked because underneath the ConvertFrom method because even though the culture is used way deep it appears is not setting any format information based on the culture. 

But this is only the tip of the iceberg because at this point we already know that the type of the converter is DoubleConverter which inherits from the BaseNumberConverter. The following is the code from the latest class I mentioned and as you can notice it would fall into the else which after doing its thing it will call the FromString method which is an abstract method that is implemented in the DoubleConverter in this particular case
       /// <devdoc>
        ///    <para>Converts the given value object to a 64-bit signed integer object.</para>
        /// </devdoc>
        public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value) {
            if (value is string) {
                string text = ((string)value).Trim();
 
                try {
                    if (AllowHex && text[0] == '#') {
                        return FromString(text.Substring(1), 16);
                    }
                    else if (AllowHex && text.StartsWith("0x") 
                             || text.StartsWith("0X")
                             || text.StartsWith("&h")
                             || text.StartsWith("&H")) {
                        return FromString(text.Substring(2), 16);
                    }
                    else {
                        if (culture == null) {
                            culture = CultureInfo.CurrentCulture;
                        }
                        NumberFormatInfo formatInfo = (NumberFormatInfo)culture.GetFormat(typeof(NumberFormatInfo));
                        return FromString(text, formatInfo);
                    }
                }
                catch (Exception e) {
                    throw FromStringError(text, e);
                }
            }
            return base.ConvertFrom(context, culture, value);
        }
In the DoubleConverter this code will execute based on the parameters
      /// <devdoc>
        /// Convert the given value to a string using the given formatInfo
        /// </devdoc>
        internal override object FromString(string value, NumberFormatInfo formatInfo) {
                return Double.Parse(value, NumberStyles.Float, formatInfo);
        }
I just made a little change in the code adding the InvariantCulture to see which method was being called and this is the inner exception:
   at System.Number.ParseDouble(String value, NumberStyles options, NumberFormatInfo numfmt)
   at System.ComponentModel.DoubleConverter.FromString(String value, NumberFormatInfo formatInfo)
   at System.ComponentModel.BaseNumberConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
So to me looks like the NumberFormatInfo is not setting correctly the value for the DecimalSeparator and I believe everything falls here:
private const NumberStyles InvalidNumberStyles = ~(NumberStyles.AllowLeadingWhite | NumberStyles.AllowTrailingWhite
                                                           | NumberStyles.AllowLeadingSign | NumberStyles.AllowTrailingSign
                                                           | NumberStyles.AllowParentheses | NumberStyles.AllowDecimalPoint
                                                           | NumberStyles.AllowThousands | NumberStyles.AllowExponent
                                                           | NumberStyles.AllowCurrencySymbol | NumberStyles.AllowHexSpecifier);
       internal static void ValidateParseStyleFloatingPoint(NumberStyles style) {
            // Check for undefined flags
            if ((style & InvalidNumberStyles) != 0) {
                throw new ArgumentException(Environment.GetResourceString("Argument_InvalidNumberStyles"), "style");
            }
            Contract.EndContractBlock();
            if ((style & NumberStyles.AllowHexSpecifier) != 0) { // Check for hex number
                throw new ArgumentException(Environment.GetResourceString("Arg_HexStyleNotSupported"));
            }
        }
    } // NumberFormatInfo
So that's why the suggested workaround which shouldn't impact anything in the config files because we are only changing it to allow the conversion. But I agree with you, I'm really no happy with this workaround but based on all my research it appears that's the only thing we could do, but I'm open for suggestions and actually you can test this very easily by doing double.parse and passing whichever culture you want to use and you will see that it fails if you pass the value with comma as separator. Anyway, let me know what do you want to do with this.
[BaseNumberConverter.cs](https://referencesource.microsoft.com/#System/compmod/system/componentmodel/basenumberconverter.cs,31448b682b8a2f02)
[NumberFormatInfo.cs](https://referencesource.microsoft.com/#mscorlib/system/globalization/numberformatinfo.cs,0ce8c4c6ad82293f)

Reply:
[I updated your comment [to use code fences - triple-backtick - so the code snippets are readable](https://help.github.com/en/articles/creating-and-highlighting-code-blocks). I hope that's OK.]

I'm a bit short on time at the moment, so I can't run my own tests or dive in here with concrete guidance. Maybe more "what I think should happen" would help:

| If the XML file has... | It should be parsed as... |
| --- | --- |
| `"4.1"` | Floating point number four-point-one: `4.1` |
| `"4,1"` |  Integer forty-one: `41` |
| `"4.100"` | Floating point number four-point-one: `4.1` |
| `"4,100"` | Integer four-thousand, one hundred: `4100` |
| `"4.100,123"` | **EXCEPTION** This isn't supported by invariant culture |

That should be the result regardless of what the culture of the current/running machine/process/thread is.

To test that, I imagine there's an XUnit test that will require setting the culture to something that uses `,` as a decimal separator - this could be done in a try/finally or using something like [this sample attribute](https://github.com/xunit/samples.xunit/blob/master/UseCulture/UseCultureAttribute.cs)

How about this:

- Update the min/max values in your tests to have four values - one for each _successful_ row I noted in the table. This will test a bunch of parsing combinations. Doesn't have to be those exact values, but the idea is to use the random formatting there to provide some challenges beyond just decimal.
- Add a test that forces the culture to `en-US` - US uses . for decimal and , for thousands. Parse the values and they should work.
- Add a test that forces the culture to `it-IT` - Italian uses , for decimal and . for thousands. Parse the values and ensure they are the same as `en-US`.

It could look something like this (off the top of my head, likely won't compile):

```c#
[InlineData("en-US")]
[InlineData("it-IT")]
[Theory]
public void GetInvariantCulture_ParametersXml(string culture)
{
    var originalCulture = Thread.CurrentThread.CurrentCulture;
    var originalUICulture = Thread.CurrentThread.CurrentUICulture;
    Thread.CurrentThread.CurrentCulture = culture;
    Thread.CurrentThread.CurrentUICulture = culture;
    CultureInfo.CurrentCulture.ClearCachedData();
    CultureInfo.CurrentUICulture.ClearCachedData();
    try
    {
      // Parse the values and assert. This should work for any of the cultures.
    }
    finally
    {
      Thread.CurrentThread.CurrentCulture = originalCulture;
      Thread.CurrentThread.CurrentUICulture = originalUICulture;
      CultureInfo.CurrentCulture.ClearCachedData();
      CultureInfo.CurrentUICulture.ClearCachedData();
    }
}
```

If the string replacement thing still works, awesome. Give it a shot with just passing InvariantCulture into the `ConvertFrom` just to see. Ideally it'd just work that way, but, again, if not and we have a test to exercise that... sounds good.

Reply:
I'll give it a try and let you know